http-server/src/server.rs

```diff
@@ -24,288 +24,202 @@
-	pub fn local_addr(&self) -> &SocketAddr {
-		&self.local_addr
+impl Default for Builder {
+	fn default() -> Self {
+		Self { max_request_body_size: 10 * 1024 * 1024, access_control: AccessControl::default(), keep_alive: true }
 	}
+}
 
-	/// Registers a notification method name towards the server.
-	///
-	/// Clients will then be able to call this method.
-	/// The returned object allows you to process incoming notifications.
-	///
-	/// If `allow_losses` is true, then the server is allowed to drop notifications if the
-	/// notifications handler (i.e. the code that uses [`RegisteredNotifications`]) is too slow
-	/// to process notifications.
-	///
-	/// Returns an error if the method name was already registered.
-	pub fn register_notification(
-		&self,
-		method_name: String,
-		allow_losses: bool,
-	) -> Result<RegisteredNotification, Error> {
-		if !self.registered_methods.lock().insert(method_name.clone()) {
-			return Err(Error::MethodAlreadyRegistered(method_name));
-		}
-
-		log::trace!("[frontend]: register_notification={}", method_name);
-		let (tx, rx) = mpsc::channel(32);
+pub struct Server {
+	/// Hyper server.
+	listener: HyperBuilder<AddrIncoming>,
+	/// Registered methods.
+	root: RpcModule,
+	/// Max request body size.
+	max_request_body_size: u32,
+	/// Access control
+	access_control: AccessControl,
+}
 
-		self.to_back
-			.unbounded_send(FrontToBack::RegisterNotifications { name: method_name, handler: tx, allow_losses })
-			.map_err(|e| Error::Internal(e.into_send_error()))?;
+impl Server {
+	/// Register a new RPC method, which responds with a given callback.
+	pub fn register_method<F, R>(&mut self, method_name: &'static str, callback: F) -> Result<(), Error>
+	where
+		R: Serialize,
+		F: Fn(RpcParams) -> Result<R, RpcError> + Send + Sync + 'static,
+	{
+		self.root.register_method(method_name, callback)
+	}
 
-		Ok(RegisteredNotification { queries_rx: rx })
+	/// Register all methods from a module on this server.
+	pub fn register_module(&mut self, module: RpcModule) -> Result<(), Error> {
+		self.root.merge(module)
 	}
 
-	/// Registers a method towards the server.
-	///
-	/// Clients will then be able to call this method.
-	/// The returned object allows you to handle incoming requests.
-	///
-	/// Contrary to [`register_notifications`](Server::register_notifications), there is no
-	/// `allow_losses` parameter here. If the handler is too slow to process requests, then the
-	/// server automatically returns an "internal error" to the client.
-	///
-	/// Returns an error if the method name was already registered.
-	pub fn register_method(&self, method_name: String) -> Result<RegisteredMethod, Error> {
-		if !self.registered_methods.lock().insert(method_name.clone()) {
-			return Err(Error::MethodAlreadyRegistered(method_name));
-		}
+	/// Start responding to connections requests. This will block current thread until the server is stopped.
+	pub async fn start(self) -> anyhow::Result<SocketAddr> {
```

Comment:
The API here is a little bit annoying because I can't get the local address of the server until `serve` is called.

I guess we could create the socket manually instead from `socket2` use `hyper::from_tcp` instead